ledger-core/vanilla/cachekit/core.go

```diff
@@ -0,0 +1,376 @@
+		count = p.rarest.trimRarest(p, count)
+		if len(p.rarest.access) == 0 {
+			p.rarest = p.rarest.next
+			p.oldest = p.rarest
+		}
+	}
+}
+
+func (p *Core) trimGenerations(count int) {
+	prev := p.oldest
+	if prev == p.rarest || prev.next != p.rarest {
+		prev = nil
+	}
+
+	for ;count > 0 && p.recent != p.rarest; count-- {
+		p.rarest.trimGeneration(p, prev)
+
+		if len(p.rarest.access) == 0 {
+			p.rarest = p.rarest.next
+			if prev != nil {
+				prev.next = p.rarest
+			}
+		} else {
+			prev = p.rarest
+			p.rarest = p.rarest.next
+		}
+	}
+}
+
+func (p *Core) ResetGenerations() {
+
+}
+
+/*******************************************/
+
+type generation struct {
+	access []uint32 // Index
+	fence  map[uint32]struct{}
+	start  Age
+	end    Age
+	next   *generation
+	genNo  GenNo
+}
+
+func (p *generation) init(capacity int, loadFence bool) {
+	p.access = make([]uint32, 0, capacity)
+	if loadFence {
+		p.fence = make(map[uint32]struct{}, capacity)
+	}
+}
+
+func (p *generation) Add(index Index, age Age) GenNo {
+	if p.end < age {
+		p.end = age
+	}
+
+	idx := uint32(index)
+	if p.fence != nil {
+		if _, ok := p.fence[idx]; !ok {
+			p.fence[idx] = struct{}{}
+			p.access = append(p.access, idx)
+		}
+	} else {
+		if n := len(p.access); n == 0 || p.access[n - 1] != idx {
+			p.access = append(p.access, idx)
+		}
+	}
+
+	return p.genNo
+}
+
+func (p *generation) trimOldest(c *Core, count int) int {
+	indices := p.access
+	i, j := 0, 0
+	for _, idx := range p.access {
+		i++
+		if v, ok := c.alloc.Get(Index(idx)); !ok || v > 0 {
+			continue
+		}
+		c.alloc.Delete(Index(idx))
+
+		indices[j] = idx
+		j++
+
+		count--
+		if count == 0 {
+			break
+		}
+	}
+
+	p.access = p.access[i:]
+	if j > 0 {
+		c.trimFn(indices[:j:j])
+	}
+
+	return count
+}
+
+func (p *generation) trimRarest(c *Core, count int) int {
+	indices := p.access
+	i, j := 0, 0
+	for _, idx := range p.access {
+		i++
+		switch v, changed, ok := c.alloc.Dec(Index(idx)); {
+		case !ok:
+			// was removed
+			continue
+		case v > 0:
+			// not yet rare
+			continue
+		case !changed:
+			// is already among the oldest
+			continue
+		}
+		c.alloc.Delete(Index(idx))
+
+		indices[j] = idx
+		j++
+		count--
+		if count == 0 {
+			break
+		}
+	}
+
+	p.access = p.access[i:]
+	if j > 0 {
+		c.trimFn(indices[:j:j])
+	}
+
+	return count
+}
+
+func (p *generation) trimGeneration(c *Core, prev *generation) {
+	indices := p.access
+	j := 0
+	for _, idx := range p.access {
+		switch v, changed, ok := c.alloc.Dec(Index(idx)); {
+		case !ok:
+			// was removed
+			continue
+		case v > 0:
+			// not yet rare
+			continue
+		case !changed:
+			// is already among the oldest
+			continue
+		}
+
+		indices[j] = idx
+		j++
+	}
+
+	if j == 0 {
+		p.access = nil
+		return
+	}
+	p.access = p.access[:j]
+	if prev == nil {
+		return
+	}
+
+	switch prevN := len(prev.access); {
+	case prevN == 0:
+		return
+	case prevN < j && prevN + j <= cap(p.access):
+		p.access = append(p.access, prev.access...)
```

Comment:
I don't see `start/end` get updated on merges, looks like bug.

Reply:
good catch

Reply:
there is also event ordering is broken here
fixed